lib/thrift/include/thrift/transport/TSSLSocket.h

```diff
@@ -0,0 +1,428 @@
+// Put this first to avoid WIN32 build failure
+#include <thrift/transport/TSocket.h>
+
+// #include <openssl/ssl.h>
+#include <string>
+#include <thrift/concurrency/Mutex.h>
+
+#include <zephyr/net/tls_credentials.h>
+#include <zephyr/posix/sys/socket.h>
+
+namespace apache {
+namespace thrift {
+namespace transport {
+
+class AccessManager;
+#if 0
+class SSLContext;
+#endif
+
+enum SSLProtocol {
+  SSLTLS = 0,  // Supports SSLv2 and SSLv3 handshake but only negotiates at TLSv1_0 or later.
+               // SSLv2   = 1,  // HORRIBLY INSECURE!
+  SSLv3 = 2,   // Supports SSLv3 only - also horribly insecure!
+  TLSv1_0 = 3, // Supports TLSv1_0 or later.
+  TLSv1_1 = 4, // Supports TLSv1_1 or later.
+  TLSv1_2 = 5, // Supports TLSv1_2 or later.
+  LATEST = TLSv1_2
+};
+
+enum ThriftTLScertificateType {
+  Thrift_TLS_CA_CERT_TAG,
+  Thrift_TLS_SERVER_CERT_TAG,
+  Thrift_TLS_PRIVATE_KEY,
+};
+
+#define TSSL_EINTR 0
+#define TSSL_DATA 1
+
+/**
+ * Initialize OpenSSL library.  This function, or some other
+ * equivalent function to initialize OpenSSL, must be called before
+ * TSSLSocket is used.  If you set TSSLSocketFactory to use manual
+ * OpenSSL initialization, you should call this function or otherwise
+ * ensure OpenSSL is initialized yourself.
+ */
+void initializeOpenSSL();
+/**
+ * Cleanup OpenSSL library.  This function should be called to clean
+ * up OpenSSL after use of OpenSSL functionality is finished.  If you
+ * set TSSLSocketFactory to use manual OpenSSL initialization, you
+ * should call this function yourself or ensure that whatever
+ * initialized OpenSSL cleans it up too.
+ */
+void cleanupOpenSSL();
+
+/**
+ * OpenSSL implementation for SSL socket interface.
+ */
+class TSSLSocket : public TSocket {
+public:
+  ~TSSLSocket() override;
+/**
+ * TTransport interface.
+ */
+#if 0
+  bool isOpen() const override;
+  bool peek() override;
+  void open() override;
+  void close() override;
+  bool hasPendingDataToRead() override;
+  uint32_t read(uint8_t* buf, uint32_t len) override;
+  void write(const uint8_t* buf, uint32_t len) override;
+  uint32_t write_partial(const uint8_t* buf, uint32_t len) override;
+  void flush() override;
+#endif
+  void open() override;
+  /**
+   * Set whether to use client or server side SSL handshake protocol.
+   *
+   * @param flag  Use server side handshake protocol if true.
+   */
+  void server(bool flag) { server_ = flag; }
+  /**
+   * Determine whether the SSL socket is server or client mode.
+   */
+  bool server() const { return server_; }
+  /**
+   * Set AccessManager.
+   *
+   * @param manager  Instance of AccessManager
+   */
+  virtual void access(std::shared_ptr<AccessManager> manager) { access_ = manager; }
+  /**
+   * Set eventSafe flag if libevent is used.
+   */
+  void setLibeventSafe() { eventSafe_ = true; }
+  /**
+   * Determines whether SSL Socket is libevent safe or not.
+   */
+  bool isLibeventSafe() const { return eventSafe_; }
+
+  void authenticate(bool required);
```

Comment:
Ah, I see - please undelete the following methods:
```
-  /**
-   * Authorize peer access after SSL handshake completes.
-   */
-  virtual void authorize();
-  /**
-   * Initiate SSL handshake if not already initiated.
-   */
-  void initializeHandshake();
-  /**
-   * Initiate SSL handshake params.
-   */
-  void initializeHandshakeParams();
-  /**
-   * Check if  SSL handshake is completed or not.
-   */
-  bool checkHandshake();
-  /**
-   * Waits for an socket or shutdown event.
-   *
-   * @throw TTransportException::INTERRUPTED if interrupted is signaled.
-   *
-   * @return TSSL_EINTR if EINTR happened on the underlying socket
-   *         TSSL_DATA  if data is available on the socket.
-   */
-  unsigned int waitForEvent(bool wantRead);
```

Even if they do nothing, it's fine. I guess the only tricky one would be `waitForEvent()`, but it might be possible to use `poll(2)` to achieve what you need.

Reply:
The zephyr way of setting authenticate mode requires the socket is set up first. However, the thrift API made this setting exposed by the factory of TSSLSocket, which exists before any TSSLSocket is created. Therefore, the factory has to remember the authenticate mode set by the user, and pass it to the TSSLSocket it creates. As the factory is set as a friend class of TSSLSocket, it can directly set the mode member of TSSLSocket, but I decided to make a function for this nevertheless. It can be removed and the factory have to directly set the field of TSSLSocket creates.

Reply:
Also note, that this was removed, which might actually be kind of useful for us.
```
-  std::shared_ptr<SSLContext> ctx_;
```